feat: add configuration for user private assistants

leon-juenemann opened the pull request:
## Summary

Adds the option to only see assistants that you (the user) created or that don't have an author. This is important for privacy reasons when multiple users are sharing the same API-Key. Otherwise a user could get personal information from e.g. a file that another user uploaded.
See this isssue: https://github.com/danny-avila/LibreChat/issues/1926

The specific use case that led to this PR:
- we don't want users to see other users assistants
- we still want to make some assistants available to all users (assistants without a metadata.author field)
## Change Type

Please delete any irrelevant options.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Testing
Note: assistant configuration for azure is currently not working as intended (see https://github.com/danny-avila/LibreChat/pull/2824/files)
In librechat.yaml:
```
endpoints:
  assistants:
    userOnly: true
```
1. Login User1 and create an assistant.
2. Login User2 and does not have access to the assistant.
### **Test Configuration**:

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] Local unit tests pass with my changes
- [x] A pull request for updating the documentation has been submitted. (https://github.com/LibreChat-AI/librechat.ai/pull/37)

leon-juenemann-mw committed:
feat: add configuration for user private assistants

danny-avila commented:
Thanks! you should also add a check when users make message requests, see where else `excludedIds` is used

danny-avila commented:
As test failure shows, `private` is a reserved keyword in JavaScript so we need to name the field differently. I think `privateAssistants` suffices

Also adding a test and updating `api/server/services/AppService.spec.js` would be great, too.

leon-juenemann-mw committed:
filter private assistant message requests

leon-juenemann commented:
Thank you for your quick feedback.
1. renamed 'private' -> 'privateAssistants'
2. add check for private assistants when users make message

I had some problems accessing the assistant author (metadata) in the message request.
Did some light refactoring but I'm not super happy with my solution. See my comment in code.

leon-juenemann commented on api/server/controllers/assistants/chatV1.js:
```diff
@@ -60,28 +59,18 @@ const chatV1 = async (req, res) => {
-
-  if (assistantsConfig) {
-    const { supportedIds, excludedIds } = assistantsConfig;
+  req.query.endpoint = endpoint;
```
Not happy with this line. I struggled to get the correct request object to pass into fetchAssistants(). Maybe there is another way to get the assistants?

leon-juenemann-mw committed:
add test for privateAssistants

leon-juenemann-mw committed:
add privateAssistants configuration to tests

leon-juenemann commented:
I added and adjusted tests in `api/server/services/AppService.spec.js`.

What needs to happen for this PR to be merged?
1.  Do I need to rebase to include most recent changes?
2. Create the PR for updating the documentation? I want to wait for this until you are fine with this PR

danny-avila commented:
> What needs to happen for this PR to be merged?
> 
>     1. Do I need to rebase to include most recent changes?

I can take over from here, there is a simpler way to do what you're doing with fetching the assistants to get the metadata. in fact, we can avoid the api call altogether (ideal) and only do it as a fallback. Also, you only updated v1 which is only used for azure.

>     2. Create the PR for updating the documentation? I want to wait for this until you are fine with this PR

You can feel free to do this now. it would need to mark a version bump for the librechat.yaml config

danny-avila commented:
Was hoping to get to this today along with a few other PRs but ran out of time. Will try to be back on it Sunday/Monday

DenisPalnitsky commented:
Thanks for looking into it. That's as useful feature to have 👍

danny-avila commented:
Continuing this in https://github.com/danny-avila/LibreChat/pull/2881
